src/casts/specify-representation.rkt

```diff
@@ -1236,90 +1213,112 @@ but a static single assignment is implicitly maintained.
-               (sr-tagged-array-ref t2 TYPE-TUPLE-TAG TYPE-TUPLE-COUNT-INDEX))
-             (App-Code cast-tuple (list v t1 t2 lbl ZERO-IMDT tagged-count))))
-         (let ([cast-tuple (get-cast-tuple! uid)])
-           (if (Var? v)
-               (invoke-cast-tuple cast-tuple v t1 t2 lbl)
-               (begin$
-                 (assign$ tuple-val1 v)
-                 (invoke-cast-tuple cast-tuple tuple-val1 t1 t2 lbl))))]
+         (define cast-tuple (get-cast-tuple-in-place! cast))
+         (: invoke-cast-tuple ((Var Uid) (Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-cast-tuple v t1 t2)
+           (app-code$ cast-tuple v t1 t2 l mono-address))
+         (cond
+           [(and (Var? v) (Var? t1) (Var? t2)) (invoke-cast-tuple v t1 t2)]
+           [(and (Var? v) (Var? t1))
+            (begin$ (assign$ a-t2 t2) (invoke-cast-tuple v t1 a-t2))]
+           [(and (Var? v) (Var? t2))
+            (begin$ (assign$ a-t1 t1) (invoke-cast-tuple v a-t1 t2))]
+           [(and (Var? t1) (Var? t2))
+            (begin$ (assign$ a-v v) (invoke-cast-tuple a-v t1 t2))]
+           [(and (Var? v))
+            (begin$
+              (assign$ a-t1 t1) (assign$ a-t2 t2) (invoke-cast-tuple v a-t1 a-t2))]
+           [(and (Var? t2))
+            (begin$
+              (assign$ a-v v) (assign$ a-t1 t1) (invoke-cast-tuple a-v a-t1 t2))]
+           [(and (Var? t1))
+            (begin$
+              (assign$ a-v v) (assign$ a-t2 t2) (invoke-cast-tuple a-v t1 a-t2))]
+           [else (begin$
+                   (assign$ a-v v)
+                   (assign$ a-t1 t1)
+                   (assign$ a-t2 t2)
+                   (invoke-cast-tuple a-v a-t1 a-t2))])]
+        [(Cast-Tuple cast (app recur v) (app recur t1) (app recur t2) (app recur l))
+         (define cast-tuple (get-cast-tuple! cast))
+         (: invoke-cast-tuple ((Var Uid) (Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-cast-tuple v t1 t2) (app-code$ cast-tuple v t1 t2 l))
+         (cond
+           [(and (Var? v) (Var? t1) (Var? t2)) (invoke-cast-tuple v t1 t2)]
+           [(and (Var? v) (Var? t1))
+            (begin$ (assign$ a-t2 t2) (invoke-cast-tuple v t1 a-t2))]
+           [(and (Var? v) (Var? t2))
+            (begin$ (assign$ a-t1 t1) (invoke-cast-tuple v a-t1 t2))]
+           [(and (Var? t1) (Var? t2))
+            (begin$ (assign$ a-v v) (invoke-cast-tuple a-v t1 t2))]
+           [(and (Var? v))
+            (begin$
+              (assign$ a-t1 t1) (assign$ a-t2 t2) (invoke-cast-tuple v a-t1 a-t2))]
+           [(and (Var? t2))
+            (begin$
+              (assign$ a-v v) (assign$ a-t1 t1) (invoke-cast-tuple a-v a-t1 t2))]
+           [(and (Var? t1))
+            (begin$
+              (assign$ a-v v) (assign$ a-t2 t2) (invoke-cast-tuple a-v t1 a-t2))]
+           [else (begin$
+                   (assign$ a-v v)
+                   (assign$ a-t1 t1)
+                   (assign$ a-t2 t2)
+                   (invoke-cast-tuple a-v a-t1 a-t2))])]
```

Comment:
This chunk of code, that is doing manual ANF, doesn't seem to be necessary.

Reply:
Done.